plugins/eos/gs-plugin-eos.c

```diff
@@ -267,6 +310,330 @@ read_icon_replacement_overrides (GHashTable *replacement_app_lookup)
+			gs_plugin_event_set_app (event, app);
+			gs_plugin_event_set_error (event, local_error);
+			gs_plugin_event_add_flag (event, GS_PLUGIN_EVENT_FLAG_WARNING);
+			gs_plugin_report_event (plugin, event);
+		}
+
+		break;
+	}
+	case EOS_UPDATER_STATE_FETCHING: {
+		guint64 total_size = 0;
+		guint64 downloaded = 0;
+		gfloat progress = 0;
+
+		app_ensure_installing_state (app);
+
+		downloaded = eos_updater_get_downloaded_bytes (priv->updater_proxy);
+		total_size = eos_updater_get_download_size (priv->updater_proxy);
+
+		if (total_size == 0) {
+			g_debug ("OS upgrade %s total size is 0!",
+				 gs_app_get_unique_id (app));
+		} else {
+			/* set progress only up to a max percentage, leaving the
+			 * remaining for applying the update */
+			progress = (gfloat) downloaded / (gfloat) total_size *
+				   (gfloat) max_progress_for_update;
+		}
+		gs_app_set_progress (app, (guint) progress);
+
+		break;
+	}
+	case EOS_UPDATER_STATE_UPDATE_READY: {
+		/* if there's an update ready to deployed, and it was started by
+		 * the user, we should proceed to applying the upgrade */
+		if (os_upgrade_get_download_by_user (app)) {
+			app_ensure_installing_state (app);
+			gs_app_set_progress (app, max_progress_for_update);
+			eos_updater_call_apply (priv->updater_proxy, NULL, NULL,
+						NULL);
+		} else {
+			/* otherwise just show it as available so the user has a
+			 * chance to click 'download' which deploys the update */
+			gs_app_set_state (app, AS_APP_STATE_AVAILABLE);
+		}
+
+		break;
+	}
+	case EOS_UPDATER_STATE_APPLYING_UPDATE: {
+		/* set as 'installing' because if it is applying the update, we
+		 * want to show the progress bar */
+		app_ensure_installing_state (app);
+
+		/* set up the fake progress to inform the user that something
+		 * is still being done (we don't get progress reports from
+		 * deploying updates) */
+		if (priv->upgrade_fake_progress_handler != 0)
+			g_source_remove (priv->upgrade_fake_progress_handler);
+		priv->upgrade_fake_progress = 0;
+		priv->upgrade_fake_progress_handler =
+			g_timeout_add ((guint) (1000.0 * EOS_UPGRADE_APPLY_STEP_TIME),
+				       (GSourceFunc) fake_os_upgrade_progress,
+				       plugin);
+
+		break;
+	}
+	case EOS_UPDATER_STATE_UPDATE_APPLIED: {
+		gs_app_set_state (app, AS_APP_STATE_UPDATABLE);
+
+		break;
+	}
+	default:
+		break;
+	}
+
+	current_app_state = gs_app_get_state (app);
+
+	/* reset the 'download-by-user' state if the the app is no longer
+	 * shown as downloading */
+	if (current_app_state != AS_APP_STATE_INSTALLING) {
+		os_upgrade_set_download_by_user (app, FALSE);
+	} else {
+		/* otherwise, ensure we have the right cancellable*/
+		setup_os_upgrade_cancellable (plugin);
+	}
+
+	/* if the state changed from or to 'unknown', we need to notify that a
+	 * new update should be shown */
+	if ((previous_app_state == AS_APP_STATE_UNKNOWN ||
+	     current_app_state == AS_APP_STATE_UNKNOWN) &&
+	    previous_app_state != current_app_state)
```

Comment:
I wonder if the last line is really required; I don't think it's possible to get all the way here with the state unchanged and still UNKNOWN

Reply:
If it's an initial state (app state = `UNKNOWN`) **and** the EOS updater reports state `NONE`, `READY`, or `POLLING` (without user intervention), the state of the app will not change, and we don't want the updates to be gathered again. Without it, since `sync_state_from_updater` is also called directly in `gs_plugin_add_distro_upgrades` (not only in the state-change callback), we could reach a state where it's calls the check for updates sequentially. It's not a likely thing to happen, but still it's good to have this safeguard IMO.